monix-execution/jvm/src/main/scala/monix/execution/misc/Local.scala

```diff
@@ -0,0 +1,104 @@
+package monix.execution.misc
+
+import java.util.UUID
+
+object Local {
+
+  type Context = scala.collection.immutable.Map[String, Option[LocalContext]]
+
+  trait LocalContext
+
+  private[this] def context(k: String, v: Option[LocalContext]): Context =
+    scala.collection.immutable.Map(k -> v)
+
+  private[this] val localContext = ThreadLocal[Context]()
+
+  private def register(): String = synchronized {
+    UUID.randomUUID().toString
+  }
+
+  /**
+    * Return the state of the current Local state.
+    */
+  def getContext(): Context =
+    localContext.get
+
+  /**
+    * Restore the Local state to a given Context.
+    */
+  def setContext(ctx: Context): Unit =
+    localContext.set(ctx)
+
+  /**
+    * Clear the Local state.
+    */
+  def clearContext(): Unit =
+    localContext.set(null)
+
+  /**
+    * Execute a block of code using the specified state of Local
+    * Context and restore the current state when complete
+    */
+  def withContext[T](ctx: Context)(f: => T): T = {
+    val saved = getContext()
+    setContext(ctx)
+    try f
+    finally setContext(saved)
+  }
+
+  /**
+    * Execute a block of code with a clear state of Local
+    * Context and restore the current state when complete
+    */
+  def withClearContext[T](f: => T): T = withContext(null)(f)
+
+  private def save(id: String, tctx: Option[LocalContext]): Unit = {
+    val newCtx = Option(localContext.get) match {
+      case Some(ctx) =>
+        ctx + (id -> tctx)
+      case None =>
+        context(id, tctx)
+    }
+    localContext.set(newCtx)
+  }
+
+  private def get(id: String): Option[_ <: LocalContext] = {
+    Option(localContext.get).flatMap(_.get(id).flatten)
+  }
+
+  private def clear(id: String): Unit =
+    Option(localContext.get).foreach { m =>
+      localContext.set(m - id)
+    }
+
+}
+
+final class Local[T <: Local.LocalContext] {
```

Comment:
Definitely needs ScalaDoc, with samples.

Reply:
Also, we probably need a `Local` in `monix.eval` that is purely function (i.e. that exposes methods returning `cats.Eval` or `Coeval` ;-)

We'll think about that as it's extra, don't worry about it for now.